### PR TITLE
Some questions about pixman damage tracking (could optimize a bit right away).

### PR DESCRIPTION
**Pull request by Dickby**
I'm not familiar with pixman, 
since you know better probably, can you take a look at this

### CODE REVIEW

**Comment by Dickby** [comment]
And can't it be checked once for every window, before you call any of those functions?

**Comment by Dickby** [comment]
I have to work on another project for a while now.
Please take a look at this.

File: src/render/decorations/CHyprDropShadowDecoration.cpp [hunk 1]
```diff
@@ -69,6 +69,8 @@ void CHyprDropShadowDecoration::draw(CMonitor* pMonitor, float a) {
         return; // disabled
 
     // get the real offset
+    // TODO: please tell me you didn't meant to do this every frame (in a try catch block???)
```
Comment (vaxerski) [inline src/render/decorations/CHyprDropShadowDecoration.cpp:72]: classic example of me saying "I'll fix this later".

File: src/render/OpenGL.cpp [hunk 2]
```diff
@@ -320,6 +320,10 @@ void CHyprOpenGLImpl::renderRectWithDamage(wlr_box* box, const CColor& col, pixm
     RASSERT((box->width > 0 && box->height > 0), "Tried to render rect with width/height < 0!");
     RASSERT(m_RenderData.pMonitor, "Tried to render rect without begin()!");
 
+    // TODO: can we early out here?
```
Comment (vaxerski) [inline src/render/OpenGL.cpp:323]: yeah, but the damage won't be empty since we're rendering a frame.

File: src/render/OpenGL.cpp [hunk 3]
```diff
@@ -455,7 +460,8 @@ void CHyprOpenGLImpl::renderTextureInternalWithDamage(const CTexture& tex, wlr_b
         glUniform1i(shader->applyTint, 0);
     }
 
-    glVertexAttribPointer(shader->posAttrib, 2, GL_FLOAT, GL_FALSE, 0, fullVerts);
+    // TODO: ???
+    // glVertexAttribPointer(shader->posAttrib, 2, GL_FLOAT, GL_FALSE, 0, fullVerts);
```
Comment (vaxerski) [inline src/render/OpenGL.cpp:464]: dunno. That's how others do it.

File: src/render/decorations/CHyprDropShadowDecoration.cpp (hunk 1 above)
Comment (Dickby) [reply src/render/decorations/CHyprDropShadowDecoration.cpp:72]: It's easy to fix!
I'll leave that to you.

File: src/render/OpenGL.cpp [hunk 4]
```diff
@@ -691,6 +704,10 @@ void CHyprOpenGLImpl::renderTextureWithBlur(const CTexture& tex, wlr_box* pBox,
     pixman_region32_init(&damage);
     pixman_region32_intersect_rect(&damage, m_RenderData.pDamage, pBox->x, pBox->y, pBox->width, pBox->height);  // clip it to the box
 
+    // TODO(Dickby): can we early out here?
```
Comment (vaxerski) [inline src/render/OpenGL.cpp:707]: yeah, good idea

File: src/render/OpenGL.cpp [hunk 5]
```diff
@@ -777,6 +795,7 @@ void CHyprOpenGLImpl::renderBorder(wlr_box* box, const CColor& col, int round) {
     static auto *const PBORDERSIZE = &g_pConfigManager->getConfigValuePtr("general:border_size")->intValue;
     static auto *const PMULTISAMPLE = &g_pConfigManager->getConfigValuePtr("decoration:multisample_edges")->intValue;
 
+    // TODO: pixman early out here?
```
Comment (vaxerski) [inline src/render/OpenGL.cpp:798]: same as above, damage should never be empty here

File: src/render/decorations/CHyprDropShadowDecoration.cpp (hunk 1 above)
Comment (vaxerski) [reply src/render/decorations/CHyprDropShadowDecoration.cpp:72]: ik sure sec

File: src/render/OpenGL.cpp (hunk 3 above)
Comment (Dickby) [reply src/render/OpenGL.cpp:464]: Just look 10 lines further below there is an if-else block that sets it anyways

File: src/render/OpenGL.cpp (hunk 2 above)
Comment (Dickby) [reply src/render/OpenGL.cpp:323]: will it never be empty?
And if so, why are you checking it later in the function?

File: src/render/OpenGL.cpp [hunk 6]
```diff
@@ -677,6 +683,13 @@ void CHyprOpenGLImpl::preWindowPass() {
 void CHyprOpenGLImpl::renderTextureWithBlur(const CTexture& tex, wlr_box* pBox, float a, wlr_surface* pSurface, int round) {
     RASSERT(m_RenderData.pMonitor, "Tried to render texture with blur without begin()!");
 
+    // TODO: while those static const pointers to the data values look kind of clever,
+    //       it is an indirection for every config value every time for everything.
+    //       why not have it be values send to these functions and just read them?
+    //       And btw these auto keyword thing might make it easier to write, but
+    //       If you want people to contribute, it makes it more hard to read and understand for them!
+    //       I myself would use auto only if there is a real good reason for that (e.g. pointer to function that returns an array of
+    //       function pointers or something like that, or maybe unnamed structs in C)
```
[the remark this answers is not in the document]
Comment (Dickby) [reply src/render/OpenGL.cpp:692]: Looks like fucking python or something, i hate that auto thing!
The worst is, there is now the same keyword in C23 standard.
By the way, i would give so much for this project being written in C!

[the remark this answers is not in the document]
Comment (vaxerski) [reply src/render/OpenGL.cpp:692]: I would give so much for this project not being written in C

File: src/render/decorations/CHyprDropShadowDecoration.cpp (hunk 1 above)
Comment (vaxerski) [reply src/render/decorations/CHyprDropShadowDecoration.cpp:72]: done

File: src/render/OpenGL.cpp (hunk 2 above)
Comment (vaxerski) [reply src/render/OpenGL.cpp:323]: oh, dunno, maybe a leftover

File: src/render/OpenGL.cpp (hunk 3 above)
Comment (vaxerski) [reply src/render/OpenGL.cpp:464]: done

**Comment by Dickby** [comment]
I would like someone (could be me or you) to rewrite own (possible slimmer and more suited for the use case)
versions of the STL functions you use!
That could cut the compile time a lot!.
If you would consider a unity build (not that engine the design concept of pound including source files into one single
compilation unit) i'm shure we could make the whole program compile in less than 10 seconds, that would 
developing much faster!

**Comment by vaxerski** [comment]
that is true. But we should focus on one thing at a time.

File: src/render/Renderer.cpp [hunk 7]
```diff
@@ -36,18 +36,15 @@ void renderSurface(struct wlr_surface* surface, int x, int y, void* data) {
...
-        g_pHyprOpenGL->renderTexture(TEXTURE, &windowBox, RDATA->fadeAlpha * RDATA->alpha, rounding, true);
+    // TODO: this is purely cosmetic (the compiler will know to do this).
+    //       I write code like this too: adding stuff trying and end up with something that works,
+    //       but eventually somebody should look over it to make it's kind of readable (so i did!).
+    if (RDATA->surface && surface == RDATA->surface &&
+	!(wlr_surface_is_xwayland_surface(surface) && !wlr_xwayland_surface_from_wlr_surface(surface)->has_alpha && RDATA->fadeAlpha * RDATA->alpha == 255.f) &&
+         RDATA->blur) {
+	g_pHyprOpenGL->renderTextureWithBlur(TEXTURE, &windowBox, RDATA->fadeAlpha * RDATA->alpha, surface, rounding);
+    }else {
+	g_pHyprOpenGL->renderTexture(TEXTURE, &windowBox, RDATA->fadeAlpha * RDATA->alpha, rounding, true);
```
Comment (Dickby) [inline src/render/Renderer.cpp:47]: New

File: src/render/decorations/CHyprDropShadowDecoration.cpp [hunk 8]
```diff
@@ -89,7 +91,13 @@ void CHyprDropShadowDecoration::draw(CMonitor* pMonitor, float a) {
...
 
     // draw the shadow
-    wlr_box fullBox = {m_vLastWindowPos.x - m_seExtents.topLeft.x + 2, m_vLastWindowPos.y - m_seExtents.topLeft.y + 2, m_vLastWindowSize.x + m_seExtents.topLeft.x + m_seExtents.bottomRight.x - 4, m_vLastWindowSize.y + m_seExtents.topLeft.y + m_seExtents.bottomRight.y - 4};
+    // TODO: would it be ok if i break long lines like this, if i come across them?
+    //       It's so much more readable, and not everybody is working on a super wide screeen! PLEASE!
+    wlr_box fullBox = {
+	m_vLastWindowPos.x - m_seExtents.topLeft.x + 2,
+	m_vLastWindowPos.y - m_seExtents.topLeft.y + 2,
+	m_vLastWindowSize.x + m_seExtents.topLeft.x + m_seExtents.bottomRight.x - 4,
+	m_vLastWindowSize.y + m_seExtents.topLeft.y + m_seExtents.bottomRight.y - 4};
```
Comment (Dickby) [inline src/render/decorations/CHyprDropShadowDecoration.cpp:101]: I'm beggin you here

File: src/render/decorations/CHyprDropShadowDecoration.cpp [hunk 9]
```diff
@@ -89,7 +91,13 @@ void CHyprDropShadowDecoration::draw(CMonitor* pMonitor, float a) {
     m_seExtents = {{*PSHADOWSIZE + 2 - offset.x, *PSHADOWSIZE + 2 - offset.y}, {*PSHADOWSIZE + 2 + offset.x, *PSHADOWSIZE + 2 + offset.y}};
 
     // draw the shadow
-    wlr_box fullBox = {m_vLastWindowPos.x - m_seExtents.topLeft.x + 2, m_vLastWindowPos.y - m_seExtents.topLeft.y + 2, m_vLastWindowSize.x + m_seExtents.topLeft.x + m_seExtents.bottomRight.x - 4, m_vLastWindowSize.y + m_seExtents.topLeft.y + m_seExtents.bottomRight.y - 4};
+    // TODO: would it be ok if i break long lines like this, if i come across them?
+    //       It's so much more readable, and not everybody is working on a super wide screeen! PLEASE!
```
Comment (vaxerski) [inline src/render/decorations/CHyprDropShadowDecoration.cpp:95]: I don't use an ultrawide but am used to java's `FactoryFactoryFactoryFactoryFactoryFactoryFactoryFactoryFactory`

It's ugh. I don't like it a lot, but it is more readable, yeah. Feel free, I guess.

File: src/render/decorations/CHyprDropShadowDecoration.cpp (hunk 8 above)
Comment (vaxerski) [reply src/render/decorations/CHyprDropShadowDecoration.cpp:101]: commented below

File: src/render/Renderer.cpp (hunk 7 above)
Comment (vaxerski) [reply src/render/Renderer.cpp:47]: see below

File: src/render/decorations/CHyprDropShadowDecoration.cpp (hunk 9 above)
Comment (Dickby) [reply src/render/decorations/CHyprDropShadowDecoration.cpp:95]: Love you for that!!!

File: src/render/Renderer.cpp (hunk 7 above)
Comment (Dickby) [reply src/render/Renderer.cpp:47]: Not shure if you noticed but this is not just line breaking.
It was a nested if else thing that, in the end, just did to different things.

**Comment by Dickby** [comment]
@vaxerski could you maybe cherry-pick the part of that "rework rounding shader" 
that moves the `#include "Shaders.hpp"` from OpenGL.hpp to OpenGL.cpp, for now.
It's such a pain to work on Shaders!
By now, all the time i do something new i have to do it manually!

**Comment by vaxerski** [comment]
done

**Comment by Dickby** [comment]
How do you things like that so fast?
Freaking young generation coders!
How old are you btw.

**Comment by vaxerski** [comment]
19 and I have never seen a woman

**Pull request closed without merging**